draft-irtf-cfrg-vdaf.md

```diff
@@ -1597,69 +1554,83 @@ Otherwise, if the state is `Continued`, then processing continues.
             vdaf_verify_key,
             1,
-            vdaf.decode_agg_param(agg_param),
+            self.decode_agg_param(agg_param),
             nonce,
-            vdaf.decode_public_share(public_share),
-            vdaf.decode_input_share(1, input_share),
+            self.decode_public_share(public_share),
+            self.decode_input_share(1, input_share),
         )
 
-        inbound = decode_ping_pong_message(inbound_encoded)
-
-        if inbound.type != 0: # initialize
+        ([inbound_type], inbound) = front(1, inbound)
+        if inbound_type != 0:  # initialize
             return (Rejected(), None)
 
+        (el, inbound) = front(4, inbound)
+        l = from_be_bytes(el)
+        (encoded, inbound) = front(l, inbound)
+        assert len(inbound) == 0
+
         prep_shares = [
-            vdaf.decode_prep_share(prep_state, inbound.prep_share),
+            self.decode_prep_share(prep_state, encoded),
             prep_share,
         ]
-        return vdaf.ping_pong_transition(
-            agg_param,
+        return self.ping_pong_transition(
+            self.decode_agg_param(agg_param),
             prep_shares,
             prep_state,
+            0,
         )
     except:
         return (Rejected(), None)
 ~~~
 
-Procedure `ping_pong_transition()` takes in the prep shares, combines them into
+Procedure `ping_pong_transition` takes in the prep shares, combines them into
 the prep message, and computes the next prep state of the caller:
 
-~~~ pseudocode
+~~~ python
 def ping_pong_transition(
-            vdaf,
-            agg_param: AggParam,
-            prep_shares: list[PrepShare],
-            prep_state: PrepState,
-         ) -> (State, bytes):
-    prep_msg = vdaf.prep_shares_to_prep(agg_param,
+        self,
+        agg_param: AggParam,
+        prep_shares: list[PrepShare],
+        prep_state: PrepState,
+        prep_round: int) -> tuple[State, bytes]:
+    prep_msg = self.prep_shares_to_prep(agg_param,
                                         prep_shares)
-    out = vdaf.prep_next(prep_state, prep_msg)
-    if type(out) == OutShare:
-        outbound = Message.finish(vdaf.encode_prep_msg(prep_msg))
-        return (Finished(out), encode_ping_pong_message(outbound))
-    (prep_state, prep_share) = out
-    outbound = Message.continue(
-        vdaf.encode_prep_msg(prep_msg),
-        vdaf.encode_prep_share(prep_share),
-    )
-    return (
-        Continued(prep_state),
-        encode_ping_pong_message(outbound),
-    )
+    out = self.prep_next(prep_state, prep_msg)
+    if prep_round+1 == self.ROUNDS:
+        encoded = self.encode_prep_msg(prep_msg)
+        outbound = bytes()
+        outbound += byte(2)  # finish
+        outbound += to_be_bytes(len(encoded), 4)
+        outbound += encoded
+        return (Finished(out), outbound)
+    (prep_state, prep_share) = cast(
+        tuple[PrepState, PrepShare], out)
+    encoded_prep_msg = self.encode_prep_msg(prep_msg)
```

Comment:
nit: the `prep_msg` is encoded here, and in the `if` block above. we could save a bit of code if we encoded it once, before the `if`.

Reply:
Nice, done.